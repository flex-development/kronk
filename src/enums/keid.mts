/**
 * @file Enums - keid
 * @module kronk/enums/keid
 */

import type { KronkErrorId } from '@flex-development/kronk'

/**
 * Kronk error ids.
 *
 * @see {@linkcode KronkErrorId}
 *
 * @internal
 *
 * @enum {KronkErrorId}
 */
const enum keid {
  argument_after_variadic = 'kronk/argument-after-variadic',
  duplicate_option = 'kronk/duplicate-option',
  duplicate_subcommand = 'kronk/duplicate-subcommand',
  error = 'kronk/error',
  excess_arguments = 'kronk/excess-arguments',
  invalid_argument = 'kronk/invalid-argument',
  invalid_argument_syntax = 'kronk/invalid-argument-syntax',
  invalid_flags = 'kronk/invalid-flags',
  invalid_subcommand_name = 'kronk/invalid-subcommand-name',
  missing_argument = 'kronk/missing-argument',
  missing_mandatory_option = 'kronk/missing-mandatory-option',
  no_flags = 'kronk/no-flags',
  unknown_option = 'kronk/unknown-option'
}

export default keid
