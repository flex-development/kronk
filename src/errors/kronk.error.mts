/**
 * @file Errors - KronkError
 * @module kronk/errors/KronkError
 */

import keid from '#enums/keid'
import kKronkError from '#internal/k-kronk-error'
import toList from '#internal/to-list'
import type {
  EmptyString,
  ExitCode,
  KronkErrorCause,
  KronkErrorId,
  KronkErrorInfo,
  KronkErrorJson
} from '@flex-development/kronk'
import { isNIL, shake } from '@flex-development/tutils'

/**
 * A command-line error.
 *
 * @category
 *  errors
 *
 * @class
 * @extends {Error}
 */
class KronkError extends Error {
  /**
   * Additional lines to be logged with the error.
   *
   * @public
   * @instance
   * @member {string[]} additional
   */
  public additional: string[]

  /**
   * Info about the cause of the error.
   *
   * @see {@linkcode KronkErrorCause}
   *
   * @public
   * @instance
   * @override
   * @member {KronkErrorCause | null | undefined} cause
   */
  public override cause?: KronkErrorCause | null | undefined

  /**
   * The suggested exit code to use with {@linkcode process.exit}.
   *
   * @public
   * @instance
   * @member {number} code
   */
  public code: number

  /**
   * Unique id representing the error.
   *
   * @see {@linkcode KronkErrorId}
   *
   * @public
   * @instance
   * @member {KronkErrorId} id
   */
  public id: KronkErrorId

  /**
   * Create a new kronk error.
   *
   * @see {@linkcode KronkErrorInfo}
   *
   * @param {KronkErrorInfo | string} info
   *  Info about the error or a human-readable description of the error
   */
  constructor(info: KronkErrorInfo | string)

  /**
   * Create a new kronk error.
   *
   * @see {@linkcode EmptyString}
   * @see {@linkcode ExitCode}
   * @see {@linkcode KronkErrorId}
   *
   * @param {string} reason
   *  Human-readable description of the error
   * @param {EmptyString | KronkErrorId | null | undefined} [id]
   *  Unique id representing the error
   * @param {ExitCode | null | undefined} [code]
   *  Suggested exit code to use with {@linkcode process.exit}
   */
  constructor(
    reason: string,
    id?: EmptyString | KronkErrorId | null | undefined,
    code?: ExitCode | null | undefined
  )

  /**
   * Create a new kronk error.
   *
   * @see {@linkcode EmptyString}
   * @see {@linkcode ExitCode}
   * @see {@linkcode KronkErrorId}
   * @see {@linkcode KronkErrorInfo}
   *
   * @param {KronkErrorInfo | string} info
   *  Error info or human-readable description of the error
   * @param {EmptyString | KronkErrorId | null | undefined} [id]
   *  Unique id representing the error
   * @param {ExitCode | null | undefined} [code]
   *  Suggested exit code to use with {@linkcode process.exit}
   */
  constructor(
    info: KronkErrorInfo | string,
    id?: EmptyString | KronkErrorId | null | undefined,
    code?: ExitCode | null | undefined
  ) {
    if (typeof info === 'string') info = { code, id, reason: info }

    super(info.reason)
    Error.captureStackTrace(this, this.constructor)

    this.additional = info.additional ? toList(info.additional) : []
    this.cause = info.cause
    this.code = isNIL(info.code) ? 1 : +info.code
    this.id = info.id || keid.error
    this.name = 'KronkError'

    Object.defineProperty(this, kKronkError, {
      configurable: false,
      enumerable: false,
      value: true,
      writable: false
    })
  }

  /**
   * Get the error as a JSON object.
   *
   * @see {@linkcode KronkErrorJson}
   *
   * @public
   * @instance
   *
   * @return {KronkErrorJson}
   *  JSON representation of `this` error
   */
  public toJSON(): KronkErrorJson {
    return shake({
      code: this.code,
      id: this.id,
      message: this.message, // eslint-disable-next-line sort-keys
      additional: this.additional,
      cause: this.cause,
      stack: this.stack
    })
  }

  /**
   * Get the error as a human-readable string.
   *
   * @public
   * @instance
   * @override
   *
   * @return {string}
   *  String representation of `this` error
   */
  public override toString(): string {
    return `[${this.id}] ${this.message}`
  }
}

export default KronkError
