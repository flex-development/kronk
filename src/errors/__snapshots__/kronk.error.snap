// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:errors/KronkError > #toJSON > should return error as json object 1`] = `
{
  "code": 1,
  "id": "kronk/invalid-argument",
  "message": "\`Option(-m --mask <mask>)\` does not allow \`h\`",
  "additional": [
    "Choices: 'default', 'expiresHeaderFormat', 'fullDate', 'isoDate', 'isoDateTime', 'isoTime', 'isoUtcDateTime', 'longDate', 'longTime', 'mediumDate', 'mediumTime', 'paddedShortDate', 'shortDate', and 'shortTime'",
  ],
  "cause": {
    "choice": "h",
    "choices": [
      "default",
      "expiresHeaderFormat",
      "fullDate",
      "isoDate",
      "isoDateTime",
      "isoTime",
      "isoUtcDateTime",
      "longDate",
      "longTime",
      "mediumDate",
      "mediumTime",
      "paddedShortDate",
      "shortDate",
      "shortTime",
    ],
  },
  "stack": "KronkError: \`Option(-m --mask <mask>)\` does not allow \`h\`
    at \${process.cwd()}/src/errors/__tests__/kronk.error.spec.mts:121:14
    at file://\${process.cwd()}/node_modules/@vitest/runner/dist/chunk-hooks.js:155:11
    at file://\${process.cwd()}/node_modules/@vitest/runner/dist/chunk-hooks.js:752:26
    at file://\${process.cwd()}/node_modules/@vitest/runner/dist/chunk-hooks.js:1894:20
    at new Promise (<anonymous>)
    at runWithTimeout (file://\${process.cwd()}/node_modules/@vitest/runner/dist/chunk-hooks.js:1860:10)
    at runTest (file://\${process.cwd()}/node_modules/@vitest/runner/dist/chunk-hooks.js:1571:12)
    at runSuite (file://\${process.cwd()}/node_modules/@vitest/runner/dist/chunk-hooks.js:1726:8)
    at runSuite (file://\${process.cwd()}/node_modules/@vitest/runner/dist/chunk-hooks.js:1726:8)
    at runSuite (file://\${process.cwd()}/node_modules/@vitest/runner/dist/chunk-hooks.js:1726:8)",
}
`;

exports[`unit:errors/KronkError > #toString > should return error as human-readable string 1`] = `"[kronk/invalid-flags] Option(-ws, --workspace) failed due to '-ws'"`;
